### PR TITLE
added fetch_openfmri

### PR DESCRIPTION
**Pull request by MartinPerez**
Related to issue #23 The import nilearn.datasets was not compatible with the latest nilearn. I added the fetch openfmri function and improved it to be able to access any dataset that follows the convention in the future. To diminsh the manual mantainance of the links. @bthirion do you think this is a good idea? Left the original one to compare.

### CODE REVIEW

**Comment by MartinPerez** [comment]
I am adding the glob_openfmri() to return the subject function as in the other fetch cases and the corresponding test

**Comment by MartinPerez** [comment]
I realized some datasets do not enforce the tab standard appropriately on some files

**Comment by bthirion** [comment]
Thx for looking at this. 
The first thing to do now is to watch the Travis failure: I'm not sure they're related to your commits, but in any case we need to have it green.

File: nistats/datasets.py [hunk 1]
```diff
@@ -302,3 +306,216 @@ def _glob_fiac_data():
...
+
+
+def fetch_openfmri(data_dir=None, dataset_id=105, force_download=False,
+                   sub_id=1, task_id=1, model_id=1, verbose=1):
+    '''Download openfmri datasets.
+
+    Currently the openfmri website employs 6 digits for the dataset id in
+    the specification, but only 3 digits in the file name to download. So
+    we assume datasets only until 999 with download link of the form
+    http://openfmri.s3.amazonaws.com/tarballs/ds[000]{Letter}_raw{_part[0]}.
```
Comment (bthirion) [inline nistats/datasets.py]: Should be moved to a not below the Parameters.

File: nistats/datasets.py [hunk 2]
```diff
@@ -302,3 +306,216 @@ def _glob_fiac_data():
...
+
+    Parameters
+    ----------
+    data_dir: string
+        directory where data should be downloaded and unpacked
+    dataset_id: int
+        number of the dataset to download
+    '''
```
Comment (bthirion) [inline nistats/datasets.py]: Please include the other parameters into the signature.

File: nistats/datasets.py [hunk 3]
```diff
@@ -302,3 +306,216 @@ def _glob_fiac_data():
...
+            allonsets = []
+            alldurations = []
+            event_files = []
+            # This will be used to add dummy events.
+            vol = nib.load(os.path.join(dataset_dir, 'sub{0:03d}', 'BOLD',
+                                        'task{1:03d}_run001', 'bold.nii.gz')
+                           .format(sub_id, task_id)).shape[3]
+            for i in range(len(conds)):
+                names.append(conds_name[i])
+                run_id = run[run.index('run')+3:run.index('run')+6]
```
Comment (bthirion) [inline nistats/datasets.py]: Pep 8: _+_

File: nistats/datasets.py [hunk 4]
```diff
@@ -302,3 +306,216 @@ def _glob_fiac_data():
...
+        runs.sort()
+        TR = float(open(os.path.join(dataset_dir,
+                                     'scan_key.txt')).read().split(OTHER_FILE_SEP)[1])
+        _subject_data['TR'] = TR
+        _run_events = {}
+        all_event_files = []
+
+        for run in runs:
+            names = []
+            allonsets = []
```
Comment (bthirion) [inline nistats/datasets.py]: all_onsets instead of allonsets

File: nistats/datasets.py [hunk 5]
```diff
@@ -302,3 +306,216 @@ def _glob_fiac_data():
...
+            alldurations = []
```
Comment (bthirion) [inline nistats/datasets.py]: all_durations insteal of alldurations

**Comment by MartinPerez** [comment]
I will take a look into the travis failure. @bthirion please read my comment at #23 before continuing the review of the code. It might be worth waiting a bit since all the download conventions will be broken in a couple of months. Also all openfmri datasets will be migrated according to Poldrack.

File: nistats/datasets.py [hunk 6]
```diff
@@ -302,3 +306,216 @@ def _glob_fiac_data():
...
+                                            header=None)
+                    onsets = cond_info[0].tolist()
+                    durations = cond_info[1].tolist()
+                else:
+                    print 'empty file found: ' + cond_file
+                    onsets = [vol*TR - 0.1]
+                    durations = [0.0]
+                allonsets.append(onsets)
+                alldurations.append(durations)
+                n_on = len(list(itertools.chain.from_iterable(allonsets)))
```
Comment (bthirion) [inline nistats/datasets.py]: n_onsets instead of n_on

File: nistats/datasets.py [hunk 7]
```diff
@@ -302,3 +306,216 @@ def _glob_fiac_data():
...
+        except urllib2.HTTPError, e:
+            return False
+        except urllib2.URLError, e:
+            return False
+
+    def explore_posible_urls():
+        files = []
+        base_url = 'http://openfmri.s3.amazonaws.com/tarballs/ds{0:03d}{1}_raw{2}.tgz'
+        dataset_links_found = False
+        while(not dataset_links_found):
```
Comment (bthirion) [inline nistats/datasets.py]: We should avoid `while` loops

File: nistats/datasets.py [hunk 8]
```diff
@@ -302,3 +306,216 @@ def _glob_fiac_data():
...
+    def check_link_exist(link):
+        try:
+            urllib2.urlopen(link)
+            return True
+        except urllib2.HTTPError, e:
+            return False
+        except urllib2.URLError, e:
+            return False
+
+    def explore_posible_urls():
```
Comment (bthirion) [inline nistats/datasets.py]: Explain what this function is doing.

**Comment by bthirion** [comment]
On 05/10/2015 09:59, Martin Perez-Guevara wrote:
>
> I will take a look into the travis failure. @bthirion 
> <https://github.com/bthirion> please read my comment at #23 
> <https://github.com/nistats/nistats/issues/23> before continuing the 
> review of the code. It might be worth waiting a bit since all the 
> download conventions will be broken in a couple of months. Also all 
> openfmri datasets will be migrated according to Poldrack.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/nistats/nistats/pull/30#issuecomment-145456110>.
>
Sure, but I think that my comments remain relevant ;-)

File: nistats/datasets.py (hunk 8 above)
Comment (dohmatob) [inline nistats/datasets.py]: typo: _posible ==> _possible

File: nistats/datasets.py [hunk 9]
```diff
@@ -302,3 +306,216 @@ def _glob_fiac_data():
...
+    urls = [('ds{0:03d}'.format(dataset_id), f,
+             {'uncompress': True}) for f in explore_posible_urls()]
+    output_dir = os.path.join(data_dir, 'ds{0:03d}'.format(dataset_id))
+    if not os.path.exists(output_dir) and not force_download:
+        _fetch_files(data_dir, urls, verbose=verbose)
+
+    return _glob_openfmri_data()
+
+
+def fetch_openfmri2(data_dir, dataset_id, force_download=False, verbose=1):
```
Comment (dohmatob) [inline nistats/datasets.py]: Merge this with `fetch_openmri` or decide which of the two to keep.

File: examples/plot_openfmri_ds001.py [hunk 10]
```diff
@@ -0,0 +1,214 @@
+""" Example of a simple GLM fit on an OpenfMRI dataset subject
+
+The script defines a General Linear Model and fits it to OpenfMRI data.
+
```
Comment (bthirion) [inline examples/plot_openfmri_ds001.py]: Describe a bit more what's in there. Provide a pointer if possible.

Comment (bthirion) [inline examples/plot_openfmri_ds001.py]: Warn people that this is a pretty large dataset

File: examples/plot_openfmri_ds001.py [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+
+# Utility to load data from openfmri dataset for a subject
+
+# openfmri dataset extraction
+OPENFMRI_SEP = '\t'
+OTHER_FILE_SEP = ' '
+
+
+def _glob_openfmri_data(dataset_dir, sub_id, model_id, task_id):
```
Comment (bthirion) [inline examples/plot_openfmri_ds001.py]: 'sub_id' is not great naming. Should be mroe explicit, e.g. 'subject_id'

File: examples/plot_openfmri_ds001.py [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+    '''Extracts model from openfmri dataset for given task and subject'''
```
Comment (bthirion) [inline examples/plot_openfmri_ds001.py]: model -> model specification

File: examples/plot_openfmri_ds001.py [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+                          header=None)
+    info_conditions = cond_df[cond_df[0] == 'task{0:03d}'.format(task_id)]
+    conds = info_conditions[1].tolist()
+    conds_name = info_conditions[2].tolist()
+    run_path = os.path.join(dataset_dir, 'sub{0:03d}'.format(sub_id),
+                            'BOLD', 'task{0:03d}_run*'.format(task_id),
+                            'bold.nii.gz')
+    runs = glob.glob(run_path)
+    runs.sort()
+    TR = float(open(os.path.join(dataset_dir, 'scan_key.txt'))
```
Comment (bthirion) [inline examples/plot_openfmri_ds001.py]: `tr` instead of `TR`

File: examples/plot_openfmri_ds001.py [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+               .read().split(OTHER_FILE_SEP)[1])
+    _run_events = {}
+    all_event_files = []
+
+    for run in runs:
+        names = []
+        allonsets = []
```
Comment (bthirion) [inline examples/plot_openfmri_ds001.py]: all_onsets

File: examples/plot_openfmri_ds001.py [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+        alldurations = []
```
Comment (bthirion) [inline examples/plot_openfmri_ds001.py]: all_durations

File: examples/plot_openfmri_ds001.py [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+        event_files = []
+        # This will be used to add dummy events.
+        vol = nib.load(os.path.join(dataset_dir, 'sub{0:03d}', 'BOLD',
+                                    'task{1:03d}_run001', 'bold.nii.gz')
+                       .format(sub_id, task_id)).shape[3]
+        for i in range(len(conds)):
+            names.append(conds_name[i])
+            run_id = run[run.index('run')+3:run.index('run')+6]
+            cond_file = onsets_template.format(run_id, conds[i])
```
Comment (bthirion) [inline examples/plot_openfmri_ds001.py]: condition_file etc.

File: examples/plot_openfmri_ds001.py [hunk 17]
```diff
@@ -0,0 +1,214 @@
...
+    _run_events = {}
+    all_event_files = []
+
+    for run in runs:
+        names = []
+        allonsets = []
+        alldurations = []
+        event_files = []
+        # This will be used to add dummy events.
+        vol = nib.load(os.path.join(dataset_dir, 'sub{0:03d}', 'BOLD',
```
Comment (bthirion) [inline examples/plot_openfmri_ds001.py]: vol is  a bad name for the number of TRs.
Rather use Nilearn than nibabel when loading data.

File: examples/plot_openfmri_ds001.py [hunk 18]
```diff
@@ -0,0 +1,214 @@
...
+                                    'task{1:03d}_run001', 'bold.nii.gz')
+                       .format(sub_id, task_id)).shape[3]
+        for i in range(len(conds)):
+            names.append(conds_name[i])
+            run_id = run[run.index('run')+3:run.index('run')+6]
+            cond_file = onsets_template.format(run_id, conds[i])
+            event_files.append(cond_file)
+            onsets = []
+            durations = []
+            # ASSUMING WE NEED AT LEAST TWO EVENTS FOR ANY CONDITION
```
Comment (bthirion) [inline examples/plot_openfmri_ds001.py]: lower case

File: examples/plot_openfmri_ds001.py [hunk 19]
```diff
@@ -0,0 +1,214 @@
...
+            if os.stat(cond_file).st_size > 0:
+                cond_info = pd.read_csv(cond_file,
+                                        sep=OPENFMRI_SEP,
+                                        header=None)
+                onsets = cond_info[0].tolist()
+                durations = cond_info[1].tolist()
+            else:
+                print 'empty file found: ' + cond_file
+                onsets = [vol*TR - 0.1]
```
Comment (bthirion) [inline examples/plot_openfmri_ds001.py]: pep8...

File: examples/plot_openfmri_ds001.py [hunk 20]
```diff
@@ -0,0 +1,214 @@
...
+                durations = [0.0]
+            allonsets.append(onsets)
+            alldurations.append(durations)
+            n_on = len(list(itertools.chain.from_iterable(allonsets)))
```
Comment (bthirion) [inline examples/plot_openfmri_ds001.py]: n_onsets

File: examples/plot_openfmri_ds001.py [hunk 21]
```diff
@@ -0,0 +1,214 @@
...
+            n_dur = len(list(itertools.chain.from_iterable(alldurations)))
```
Comment (bthirion) [inline examples/plot_openfmri_ds001.py]: n_duration

File: examples/plot_openfmri_ds001.py [hunk 22]
```diff
@@ -0,0 +1,214 @@
...
+
+def _openfrmri_contrasts(ref_design_matrix, dataset_dir, model_id, task_id):
+    # Build basic contrast_matrix.
+    contrast_matrix = np.eye(ref_design_matrix.shape[1])
+    contrasts = dict([(column, contrast_matrix[i])
+                      for i, column in enumerate(ref_design_matrix.columns)])
+
+    # Load information from openfmri files
+    task = 'task{0:03d}'.format(task_id)
+    condkey_file = os.path.join(dataset_dir, 'models',
```
Comment (bthirion) [inline examples/plot_openfmri_ds001.py]: condition_key_file

File: examples/plot_openfmri_ds001.py [hunk 23]
```diff
@@ -0,0 +1,214 @@
...
+                                'model{0:03d}'.format(model_id),
+                                'condition_key.txt')
+    cond_df = pd.read_csv(condkey_file,
```
Comment (bthirion) [inline examples/plot_openfmri_ds001.py]: condition_definition

File: examples/plot_openfmri_ds001.py [hunk 24]
```diff
@@ -0,0 +1,214 @@
...
+    contrasts_df = pd.read_csv(contrasts_file,
+                               sep=OTHER_FILE_SEP,
+                               header=None)
+    task_contrasts = contrasts_df[contrasts_df[0] == task]
+
+    # Complete contrast specification
+    for idx, crow in task_contrasts.iterrows():
+        contrasts[crow[1]] = np.zeros(ref_design_matrix.shape[1])
+        for widx, weight in enumerate(crow[2:]):
+            contrasts[crow[1]] += weight*contrasts[conds_name[widx]]
```
Comment (bthirion) [inline examples/plot_openfmri_ds001.py]: pep8

File: examples/plot_openfmri_ds001.py [hunk 25]
```diff
@@ -0,0 +1,214 @@
...
+
+    return contrasts
+
```
Comment (bthirion) [inline examples/plot_openfmri_ds001.py]: These ancillary functions are pretty ong and messy. They should be put in a side module to clarify the example.

File: examples/plot_openfmri_ds001.py [hunk 26]
```diff
@@ -0,0 +1,214 @@
...
+# dataset parameters
+sub_id = 1
+model_id = 1
+task_id = 1
```
Comment (bthirion) [inline examples/plot_openfmri_ds001.py]: All this could be done directly in line 149.

File: examples/plot_openfmri_ds001.py [hunk 27]
```diff
@@ -0,0 +1,214 @@
...
+data = _glob_openfmri_data(dataset_dir, sub_id, model_id, task_id)
+
+# construct design matrices
+design_matrices = []
+contrasts = []
+for session in range(len(data.func_events)):
+    onsets = []
+    durations = []
+    names = []
+    n_cond = len(data.func_events[session]['conditions'])
```
Comment (bthirion) [inline examples/plot_openfmri_ds001.py]: n_conditions

File: examples/plot_openfmri_ds001.py [hunk 28]
```diff
@@ -0,0 +1,214 @@
...
+        n_events = len(data.func_events[session]['durations'][c])
+        names += [data.func_events[session]['conditions'][c]] * n_events
+    paradigm = pd.DataFrame({'onset': onsets, 'duration': durations,
+                             'name': names})
+    n_scans = nib.load(data.func[session]).get_data().shape[3]
+    frame_times = np.linspace(0, (n_scans - 1) * data.TR, n_scans)
+    drift_model = 'Cosine'
+    hrf_model = 'Canonical With Derivative'
+    design_matrix = make_design_matrix(frame_times, paradigm,
+                                       hrf_model=hrf_model,
```
Comment (bthirion) [inline examples/plot_openfmri_ds001.py]: you could specify `hrf_model` and `drift_model` only here. It's not canonical design, but examples are meant to be read easily.

File: examples/plot_openfmri_ds001.py [hunk 29]
```diff
@@ -0,0 +1,214 @@
...
+for contrast_id in contrasts[1].keys()[-6:]:
+    print("\tcontrast id: %s" % contrast_id)
+    z_map, t_map, eff_map, var_map = fmri_glm.transform(
+        [scon[contrast_id] for scon in contrasts], contrast_name=contrast_id,
+        output_z=True, output_stat=True, output_effects=True,
+        output_variance=True)
+
+    # store stat maps to disk
+    for dtype, out_map in zip(['z', 't', 'effects', 'variance'],
+                              [z_map, t_map, eff_map, var_map]):
```
Comment (bthirion) [inline examples/plot_openfmri_ds001.py]: Should write all of those ?

File: examples/plot_openfmri_ds001.py [hunk 30]
```diff
@@ -0,0 +1,214 @@
...
+    data.func, design_matrices)
+
+print("Computing contrasts ..")
+output_dir = os.path.join('dataset_download', 'ds001_analysis',
+                          'sub{0:03d}'.format(sub_id))
+if not os.path.exists(output_dir):
+    os.makedirs(output_dir)
+
+anat_img = nib.load(data['anat'])
+for contrast_id in contrasts[1].keys()[-6:]:
```
Comment (bthirion) [inline examples/plot_openfmri_ds001.py]: This is obviously not the best selection of contrasts that can be made on these data.

File: nistats/datasets.py [hunk 31]
```diff
@@ -303,3 +338,79 @@ def _glob_fiac_data():
...
+        }
+
+    if dataset_id not in datasets:
+        raise Exception('Unknown dataset %s' % dataset_id)
+
+    base_url = 'http://openfmri.s3.amazonaws.com/tarballs/%s.tgz'
+    urls = [(dataset_id, base_url % f, {'uncompress':True}) for f in datasets[dataset_id]]
+    temp_dir = os.path.join(data_dir, '_%s' % dataset_id, dataset_id)
+    output_dir = os.path.join(data_dir, dataset_id)
+    print data_dir, urls
```
Comment (bthirion) [inline nistats/datasets.py]: beware P3 compatibility with print

File: nistats/design_matrix.py [hunk 32]
```diff
@@ -269,7 +269,7 @@ def make_design_matrix(
         The timing of the scans in seconds.
 
     paradigm : DataFrame instance, optional
-        Description of the experimental paradigm.
+        Description of the experimental paradigm. With one row per event.
```
Comment (bthirion) [inline nistats/design_matrix.py]: , with

File: nistats/tests/test_datasets.py [hunk 33]
```diff
@@ -46,3 +46,10 @@ def test_fiac():
     assert_true(isinstance(dataset.design_matrix1, _basestring))
     assert_true(isinstance(dataset.design_matrix2, _basestring))
     assert_true(isinstance(dataset.mask, _basestring))
+
+
+def test_fetch_openfmri_dataset():
+    return
+    # tests willbe changed, to include mock instances as in Nilearn.
+    # output_dir = datasets.fetch_openfmri(data_dir=tmpdir)
+    # assert_true(isinstance(output_dir, _basestring))
```
Comment (bthirion) [inline nistats/tests/test_datasets.py]: This will have to be tested at some point !

**Comment by MartinPerez** [comment]
This became obsolete with BIDS. I will post a new BIDS branch with downloaded and utilities soon.

**Pull request closed without merging**